### ISSUE TITLE
HOWTO: Validating model with covariates on every window in the validation set

### ISSUE THREAD

**Issue by owoshch** [unit8co/darts#575]
I want to validate the quality of the predictions of a separate dataset:

`series = TimeSeries.from_dataframe(dataframe, value_cols=["speed_y_vrf", "speed_x_vrf", "ang_rate_z_vrf"])
covariates_series = TimeSeries.from_dataframe(dataframe, value_cols=["steering", "throttle", "brake"])

scaler = Scaler()
series_scaled = scaler.fit_transform(series)

covariates_scaler = Scaler()
covariates_series_scaled = scaler.fit_transform(covariates_series)

train_dataset, val_dataset = series_scaled.split_before(0.5)
train_dataset_covariates, val_dataset_covariates = covariates_series_scaled.split_before(0.5)

model_cov = BlockRNNModel(model='LSTM', input_chunk_length=100, output_chunk_length=10, n_epochs=200, random_state=0)
model_cov.fit(series=[train_dataset], 
              past_covariates=[train_dataset_covariates],
              val_series=[val_dataset],
              val_past_covariates=[val_dataset_covariates],
              verbose=True)
`

I want to find a way to get the metrics for every prediction window in the validation set.
I watched an introductory video on YouTube https://youtu.be/Sx-uI-PypmU?t=988 on the .backtest method and it's still unclear how can I compare different models on every window from the validation set.

Could you please guide me on how to do it?

**Comment by hrzn** [comment]
I think you should look at the models' `historical_forecasts()` function.
Given that your validation set is the last 50% of your series, once you have trained your model you should be able to do something along those lines
```
model_cov.historical_forecasts(series=series, past_covariates=covariates_series, start=0.5, forecast_horizon=n, retrain=False)
```

This will return a `TimeSeries` of the `n`-steps ahead forecasts you would have obtained by using this model repeteadly to obtain predictions over the last 50% of your series. Checkout the documentation if you need more infos: https://unit8co.github.io/darts/generated_api/darts.models.forecasting.block_rnn_model.html#darts.models.forecasting.block_rnn_model.BlockRNNModel.historical_forecasts

**Comment by owoshch** [comment]
Thank you so much! It worked.
However, it's still not clear to me how to validate multivariate output of the model if I want to compute metrics for every variable I'm predicting. The code below would output list of metrics of each step:
`metrics = model_cov.backtest(series_scaled,
                  past_covariates=covariates_series_scaled,
                  forecast_horizon=10,
                  start=0.5,
                  stride=1,
                  retrain=False,
                  last_points_only=False,
                  metric=mse,
                  reduction=None, 
                  verbose=True)`

This would output a list like `[0.000457716367741197, 0.00046475015704051345, ...]`.
Is there any way to get metrics for each variable in the output? Say, I have 3 targets, so I expect metrics to be a list of lists:
`[[0.324, 0.543, 0.234], [0.4323, 0.1230, 0.043], ...]

**Comment by hrzn** [comment]
Sorry for the very late answer @owoshch. For the record, I think the way to go to keep per-component metrics from backtest would be something like:
```python
from darts.metrics import mse

def my_metric(actual, pred):
  return mse(actual, pred, reduction=lambda x: x)

metrics = model_cov.backtest(series_scaled, past_covariates=covariates_series_scaled, forecast_horizon=10, start=0.5, stride=1, retrain=False, last_points_only=False, metric=my_metric, reduction=None,  verbose=True)
```

You can also set `reduction=None` in the above call to get individual scores on all sub-slices.

In addition, starting with v0.23.0, calling `backtest()` or `historical_forecasts()` on list of series is now supported.
